### ISSUE TITLE
Memory leak?

### ISSUE THREAD

**Issue by jsvine** [pdfminer/pdfminer.six#580]
**Bug report**

## A description of the bug

Somewhere in `pdfminer.six` there appears to be a memory leak; even after a Python process is done handling a PDF, the memory it required is not released.

## Steps to reproduce the bug

1. `pip install memory_profiler`
2. Save this script as `memory.py`:

```python
from pdfminer.high_level import extract_text
from memory_profiler import profile
import gc

path = "samples/nonfree/nlp2004slides.pdf"

@profile
def main():
    text = extract_text(path)
    gc.collect()

main()
```
3. Run `python memory.py`

Output:

```
Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
     7     44.3 MiB     44.3 MiB           1   @profile
     8                                         def main():
     9     51.5 MiB      7.2 MiB           1       text = extract_text(path)
    10     51.5 MiB      0.0 MiB           1       gc.collect()
```

Note: The problem does not seem specific to `extract_text`; other, lower-level usages of `pdfminer.six` seem to create similar effects. Also: Passing `caching=False` to `extract_text` does not appear to reduce the memory leak.

I have spent some time trying to track down the source of apparent leak, but have not had much luck. I'll continue looking, but am raising this issue in the meantime, in case the cause is more readily apparent to others.

**Comment by jugmac00** [comment]
I just faced two consecutive outages of the system I maintain - I tracked down the reason to one PDF a user wanted to upload - my system extracts the text to index the PDF document. I can provide the document to a maintainer ( it should not go public). Basically, the PDF consists of slides - when I open the PDF with `evince` I can view it, but the pages get rendered very, very slowly.

So there is obviously something bogus.

Is anybody aware of a tool which "fixes" bogus PDF files? If not, I think I'll just screenshot the pages and create a new one.

**Comment by pietermarsman** [comment]
I can reproduce this issue. Will have a look at the PR's.

**Comment by pietermarsman** [comment]
Would love to merge to PR's related to this issue but they don't seem to solve the issue.

**Comment by Rumpelcita** [comment]
I'm experiencing a regression of the recursion error in issue https://github.com/pdfminer/pdfminer.six/issues/404 with version `20220524`. Could this be related?

> Hi, I'm still running into this issue on version `20220524` of pdfminer.six. I can't provide the pdf that causes the issue since it needs to be anonymized, but these are the LAParams we're using for pdfminer:
> 
> ```
> LAParams(
>     line_overlap= 0.5,
>     char_margin= 1.1,
>     word_margin= 0.2,
>     line_margin= 0.5,
>     boxes_flow= 0.5,
>     all_texts=True,
>         )
> ```
> 
> Traceback:
> 
> ```
> RecursionError: maximum recursion depth exceeded
> (245 additional frame(s) were not displayed)
> ...
>   File "pdfminer/layout.py", line 705, in analyze
>     super().analyze(laparams)
>   File "pdfminer/layout.py", line 439, in analyze
>     obj.analyze(laparams)
>   File "pdfminer/layout.py", line 705, in analyze
>     super().analyze(laparams)
>   File "pdfminer/layout.py", line 439, in analyze
>     obj.analyze(laparams)
>   File "pdfminer/layout.py", line 439, in analyze
>     obj.analyze(laparams)
> 
> maximum recursion depth exceeded
> ```

**Comment by dominikhait** [comment]
Is there any update on this issue?

**Comment by pietermarsman** [comment]
No update here yet. The root cause needs to be found first.

**Comment by yoavkedem1** [comment]
Hey, is there any update on this issue?

**Comment by dhdaines** [comment]
Yes, in addition to the things noted above, this is also a consequence of the way that pdfminer.six handles indirect object references, and it's really obvious here: https://github.com/pdfminer/pdfminer.six/blob/master/pdfminer/pdftypes.py#L93

Since a `PDFDocument` contains a lot of `PDFObjRef` objects and every `PDFObjRef` also contains a ref to its parent `PDFDocument`, well, you can imagine what happens.  It's a textbook example of how to make memory leaks with a reference counting system!

(shameless plug) it is one of the first things I fixed in [PLAYA-PDF](https://github.com/dhdaines/playa) which is a deep refactoring of pdfminer.six with a better API (but which also implements pdfminer.six's layout analysis algorithm)

Yes this can probably be fixed with weak references.  Having a minimal reproducible example would be good, because as noted above there are a number of places (not just `PDFObjRef`) where the problem can manifest.
